### PR TITLE
EOS-16473,EOS-17618: [DTM0] Implement persistent log

### PR DESCRIPTION
**Pull request by mehjoshi**
Code changes for EOS-16473 and EOS-17618.

### CODE REVIEW

File: be/ut/dtm0_log_ut.c [hunk 1]
```diff
@@ -19,12 +19,18 @@
  *
  */
 
+#include "dtm0/tx_desc.h"
```
Comment (ivan-alekhin) [inline be/ut/dtm0_log_ut.c:22]: Let's move it to the list of include directives

File: be/ut/dtm0_log_ut.c [hunk 2]
```diff
@@ -273,12 +279,319 @@ void test_volatile_dtm0_log(void)
...
+	M0_UT_ASSERT(rc != 0);
+
+	ut_dl_set_pa_state(&txd[0].dtd_pg.dtpg_pa[0], M0_DTPS_PERSISTENT);
+	ut_dl_set_pa_state(&txd[0].dtd_pg.dtpg_pa[1], M0_DTPS_PERSISTENT);
+	M0_SET0(&cred);
+
+	m0_be_dtm0_log_credit(M0_DTML_PERSISTENT, &txd[0], &buf[0], seg, NULL, &cred);
+	m0_be_ut_tx_init(tx, ut_be);
+	m0_be_tx_prep(tx, &cred);
+	rc = m0_be_tx_open_sync(tx);
```
Comment (ivan-alekhin) [inline be/ut/dtm0_log_ut.c:468]: `M0_UT_ASSERT(rc == 0)`?

**Review by ivan-alekhin — COMMENTED** [review]
I have posted a few minor suggestions. We can address them in a separate patch if it is needed.

File: be/dtm0_log.c [hunk 3]
```diff
@@ -38,18 +44,25 @@ M0_TL_DESCR_DEFINE(lrec, "DTM0 Log", static, struct m0_dtm0_log_rec,
...
+                   M0_BE_DTM_LOG_MAGIX);
+
+M0_BE_LIST_DEFINE(lrec, static, struct m0_dtm0_log_rec);
+
 static bool m0_be_dtm0_log__invariant(const struct m0_be_dtm0_log *log)
 {
 	return _0C(log != NULL) &&
-	       _0C(log->dl_cs != NULL) &&
-	       _0C(lrec_tlist_invariant(log->dl_tlist));
+	       _0C(log->dl_cs != NULL); // &&
```
[the remark this answers is not in the document]
Comment (mehjoshi) [reply be/dtm0_log.c:57]: Done.

File: be/dtm0_log.c [hunk 4]
```diff
@@ -79,6 +92,7 @@ M0_INTERNAL int m0_be_dtm0_log_init(struct m0_be_dtm0_log **out,
 	}
 
 	m0_mutex_init(&log->dl_lock);
+	log->dl_ispstore = !isvstore;
```
[the remark this answers is not in the document]
Comment (mehjoshi) [reply be/dtm0_log.c:95]: Done.

File: be/dtm0_log.c [hunk 5]
```diff
@@ -100,17 +114,94 @@ M0_INTERNAL void m0_be_dtm0_log_fini(struct m0_be_dtm0_log **log,
...
                                        struct m0_be_seg             *seg,
+				       struct m0_dtm0_log_rec	    *rec,
                                        struct m0_be_tx_credit       *accum)
 {
 	/*TODO:Complete implementation during persistent list implementation */
 	switch (op) {
 	case M0_DTML_CREATE:
-	case M0_DTML_SENT:
+	{
+		struct m0_be_dtm0_log *log;
```
[the remark this answers is not in the document]
Comment (mehjoshi) [reply be/dtm0_log.c:182]: Logically, what we are trying to do in this function is to allocate memory on a given segment. And so it seems proper to pass the segment here.

File: be/dtm0_log.c [hunk 6]
```diff
@@ -159,9 +252,24 @@ struct m0_dtm0_log_rec *m0_be_dtm0_log_find(struct m0_be_dtm0_log    *log,
...
-	return m0_tl_find(lrec, rec, log->dl_tlist,
-                          m0_dtm0_tid_cmp(log->dl_cs, &rec->dlr_txd.dtd_id,
+	if (log->dl_ispstore) {
+		struct m0_be_list         *dl_list = log->dl_list;
+		struct m0_dtm0_log_rec *lrec;
+		struct m0_dtm0_log_rec *result = NULL;
+
+		m0_be_list_for(lrec, dl_list, lrec) {
+			if (m0_dtm0_tid_cmp(log->dl_cs, &lrec->dlr_txd.dtd_id,
+                                          id) == M0_DTS_EQ)
```
[the remark this answers is not in the document]
Comment (mehjoshi) [reply be/dtm0_log.c:262]: Done.

File: be/ut/dtm0_log_ut.c (hunk 2 above)
Comment (mehjoshi) [reply be/ut/dtm0_log_ut.c:468]: Done.

**Comment by mehjoshi** [comment]
Merging it based on final comments from Anatoliy. Will discuss and close the Prune logic offline with Anatoliy and address it in a subsequent sprint.

**Pull request merged**